dev/com.ibm.websphere.appserver.features/visibility/public/distributedMap-2.0/io.openliberty.distributedMap-2.0.feature

```diff
@@ -0,0 +1,26 @@
+-include= ~${workspace}/cnf/resources/bnd/feature.props
```

Comment:
Instead of making a new version of distributedMap, it may be better to do auto features and pull in the right bundle depending on if servlet 5.0 is enabled or previous versions of servlet.  You could pull out the two classes that depend on servlet in com.ibm.ws.dynacache and put them in their own bundle with a different package name like com.ibm.ws.dynacache.feature and then just have that be jakarta-ized.

Reply:
This sounds like a good idea. I will try to do this.

Reply:
So I think I did this a little different than you had suggested. I didn't move anything out of the dynacache bundle. I am keeping it jakartaee'd as before.

I created two auto features, that are a singleton and different versions of the same feature. The distributedCache-1.0 feature, I removed the dynacache bundle and the private servlet feature (these will be provided by the two auto-features).

The first auto feature provides the javaee dynacache bundle and private javaee servlet feature(s).

```
-include= ~${workspace}/cnf/resources/bnd/feature.props
symbolicName=io.openliberty.autoDistributedMap-1.0
visibility=private
singleton=true
IBM-App-ForceRestart: install, \
 uninstall
Subsystem-Version: 1.0.0
IBM-Provision-Capability: osgi.identity; filter:="(&(type=osgi.subsystem.feature)(osgi.identity=com.ibm.websphere.appserver.distributedMap-1.0))"
-features=\
  com.ibm.websphere.appserver.javax.servlet-3.0; ibm.tolerates:="3.1,4.0"
-bundles=\
  com.ibm.ws.dynacache
kind=ga
edition=core

```

The second auto feature provides the jakartaee'd dynacache bundle and the private jakarta servlet-5.0 feature. Really, I am not sure that I need to depend on that feature since the auto-feature auto-provisions only when the public servlet-5.0 feature is enabled (which enables the private servlet-5.0 feature).

```
-include= ~${workspace}/cnf/resources/bnd/feature.props
symbolicName=io.openliberty.autoDistributedMap-2.0
visibility=private
singleton=true
IBM-App-ForceRestart: install, \
 uninstall
Subsystem-Version: 2.0.0
IBM-Provision-Capability: osgi.identity; filter:="(&(type=osgi.subsystem.feature)(osgi.identity=com.ibm.websphere.appserver.distributedMap-1.0))",\
                          osgi.identity; filter:="(&(type=osgi.subsystem.feature)(osgi.identity=com.ibm.websphere.appserver.servlet-5.0))"
-features=\
  io.openliberty.jakarta.servlet-5.0
-bundles=\
  io.openliberty.dynacache.internal
kind=noship
edition=full
```

Let me know what you think.

Reply:
The above did not work. I removed the javax compatible bundle and feature from the distributeMap-1.0 feature. I added a dependency on a io.openliberty.distributeMapPrivate-1.0 feature, and it tolerates version 2.0.

The private features have the dependencies on the javax (1.0) and jakarta (2.0) bundles and features.

I also added dependencies on eeCompatible feature from both features to help resolve the javax and jakarta dependencies.

Reply:
So in summary, the last comment describes the changes I made that appear to make this work.